pkg/api/norman/customization/cluster/ocne_validator.go

```diff
@@ -0,0 +1,216 @@
+	NodeOcpus             int    `json:"nodeOcpus"`
+	NodePublicKeyContents string `json:"nodePublicKeyContents"`
+	NodeShape             string `json:"nodeShape"`
+	NodeVolumeGbs         int    `json:"nodeVolumeGbs"`
+	NumControlPlaneNodes  int    `json:"numControlPlaneNodes"`
+	NumWorkerNodes        int    `json:"numWorkerNodes"`
+	OciCsiImage           string `json:"ociCsiImage"`
+	PodCidr               string `json:"podCidr"`
+	ProxyEndpoint         string `json:"proxyEndpoint"`
+	Region                string `json:"region"`
+	UseNodePvEncryption   bool   `json:"useNodePvEncryption"`
+	VcnID                 string `json:"vcnId"`
+	VerrazzanoImage       string `json:"verrazzanoImage"`
+	VerrazzanoResource    string `json:"verrazzanoResource"`
+	WorkerNodeSubnet      string `json:"workerNodeSubnet"`
+
+	net *core.VirtualNetworkClient
+}
+
+var BadRequest = httperror.ErrorCode{
+	Code:   "Bad Request",
+	Status: 400,
+}
+
+func (v *Validator) ValidateOCNE(ocneConfig map[string]interface{}) error {
+	ctx := context.TODO()
+	o, err := getOCNEConfig(ocneConfig)
+	if err != nil {
+		return err
+	}
+	if err := v.validateOCNECredentials(o); err != nil {
+		return err
+	}
+
+	errorChannel := make(chan error)
+	validators := []func(){
+		func() {
+			v.validateVCNId(ctx, errorChannel, o)
+		},
+		func() {
+			v.validateSubnet(ctx, errorChannel, o, o.ControlPlaneSubnet, "controlPlaneSubnet")
+		},
+		func() {
+			v.validateSubnet(ctx, errorChannel, o, o.LoadBalancerSubnet, "loadBalancerSubnet")
+		},
+		func() {
+			v.validateSubnet(ctx, errorChannel, o, o.WorkerNodeSubnet, "workerNodeSubnet")
+		},
+		func() {
+			v.validateVerrazzano(errorChannel, o)
+		},
+		func() {
+			v.validateOCNECluster(errorChannel, o)
+		},
+	}
+
+	for _, validatorFunc := range validators {
+		go validatorFunc()
+	}
+
+	var result error
+	for i := 0; i < len(validators); i++ {
+		err = <-errorChannel
+		if err != nil {
+			result = err
+		}
+	}
+	close(errorChannel)
+	return result
+}
+
+func getOCNEConfig(ocneConfig map[string]interface{}) (*OCNEConfig, error) {
+	b, err := json.Marshal(ocneConfig)
+	if err != nil {
+		return nil, err
+	}
+	o := &OCNEConfig{}
+	if err := json.Unmarshal(b, o); err != nil {
+		return nil, err
+	}
+	return o, nil
+}
+
+func (v *Validator) validateOCNECredentials(o *OCNEConfig) error {
+	if o.CloudCredentialID == "" {
+		return httperror.NewAPIError(BadRequest, `"cloudCredentialId" field is required.`)
+	}
+
+	// CloudCredentialID is generally in the form <namespace>:<name>
+	// if namespace is not present, assume it to be cattle-global data
+	split := strings.Split(o.CloudCredentialID, ":")
+	name := split[0]
+	namespace := "cattle-global-data"
+	if len(split) > 1 {
+		namespace = split[0]
+		name = split[1]
+	}
+
+	cc, err := v.SecretLister.Get(namespace, name)
+	if err != nil {
+		return httperror.NewAPIError(httperror.NotFound, fmt.Sprintf("invalid credentials for id: %s", o.CloudCredentialID))
```

Comment:
will credentials always exist ?

Reply:
some very weird edge cases where they might not exist

Reply:
we will just check, because error is thrown anyway. almost always will be no error